packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+    //
+    //  ghost_adjs: the adjacency information for only the second ghost layer
+    //              indexed by ghost_offsets.
+    //              
+    //
+    //  ghost_offsets: the offsets into ghost_adjs, first layer ghost LIDs 
+    //                 minus n_local are used to index this 
+    //                 datastructure. I.E. ghost_offsets(0)
+    //                 corresponds to the ghost with LID n_local
+    //
+    //  gids: a vector that holds GIDs for all vertices on this process
+    //        (includes owned, and all ghosts, indexable by LID)
+    //
+    //  rand: a vector that holds random numbers generated on GID for tie
+    //        breaking. Indexed by LID.
+    //
+    //  owners: contains the process ID for the owner of each vertex.
+    //          Indexable by LID.
+    //
+    //  mapOwnedPlusGhosts: map that can translate between GID and LID
+    //
+    //  procs_to_send: for each vertex that is copied on a remote process,
+    //                 this map contains the list of processes that have
+    //                 a copy of a given vertex. Input LID, get the list
+    //                 of PIDs that have a ghost copy of that LID.
+    //  
+    //OUTPUT ARGS:
+    //  
+    //  femv: an FEMultiVector that holds a dual view of the colors.
+    //        After this call, femv contains updated colors.
+    //
+    void twoGhostLayer(const size_t n_local, const size_t n_total,
+                       const Teuchos::ArrayView<const lno_t>& adjs,
+                       const Teuchos::ArrayView<const offset_t>& offsets,
+                       const Teuchos::ArrayView<const lno_t>& ghost_adjs,
+                       const Teuchos::ArrayView<const offset_t>& ghost_offsets,
+                       const Teuchos::RCP<femv_t>& femv,
+                       const std::vector<gno_t>& gids, 
+                       const std::vector<int>& rand,
+                       ArrayView<int> owners,
+                       RCP<const map_t> mapOwnedPlusGhosts,
+		       std::unordered_map<lno_t, std::vector<int>> procs_to_send){
+      //initialize timing variables
+      double total_time = 0.0;
+      double interior_time = 0.0;
+      double comm_time = 0.0;
+      double comp_time = 0.0;
+      double recoloring_time=0.0;
+      double conflict_detection = 0.0;
+      
+      //Number of rounds we are saving statistics for
+      //100 is a decent default.
+      const int numStatisticRecordingRounds = 100;
+      
+      //includes all ghosts, including the second layer.
+      const size_t n_ghosts = rand.size() - n_local;
+
+
+      //Get the degrees of all ghost vertices
+      //This is necessary for recoloring based on degrees,
+      //we need ghost vertex degrees for consistency.
+      std::vector<int> deg_send_cnts(comm->getSize(),0);
+      std::vector<gno_t> deg_sdispls(comm->getSize()+1,0);
+      for(int i = 0; i < owners.size(); i++){
```

Comment:
I have revised the incorrect comment at the top, and will re-name the argument ghost_owners. This will alleviate the issue of having a different owners variable elsewhere, and make it more clear that this argument is expected to only hold owner information for the ghost vertices.